PCP: /api/v1/regions optimization

ekalinin opened the pull request:
<!--
Thank you for working on PD! Please read PD's [CONTRIBUTING](https://github.com/pingcap/pd/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.
-->

### What problem does this PR solve? <!--add the issue link with summary if it exists-->

PCP #1837

### What is changed and how it works?

**`convertToAPIRegions` optimization**

- reduced the amount of allocations in `convertToAPIRegions` by making a slice of `RegionInfo`.
- added a new function - `InitRegion`. Used it in a loop to init a slice
of pointers to the `api.RegionInfo` structs without additional allocations.
- added a method `HexRegionKeyStr`. It returns a string (and not a `[]byte`).
`StartKey/EndKey` fields are `string` in the `api.RegionInfo`, and we do not need one more convertation (and allocation) from []byte to string.
- added benchmarks:
  - `BenchmarkConvertToAPIRegions`
  - `BenchmarkHexRegionKey`
  - `BenchmarkHexRegionKeyStr`

Result:

```sh
# GO111MODULE=on go test github.com/pingcap/pd/server/api -run=None -bench=BenchmarkConvertToAPIRegions -benchmem > new.api.txt
➜ benchcmp old.api.txt new.api.txt
benchmark                           old ns/op     new ns/op     delta
BenchmarkConvertToAPIRegions-12     2880085       2208890       -23.30%

benchmark                           old allocs    new allocs    delta
BenchmarkConvertToAPIRegions-12     82182         52183         -36.50%

benchmark                           old bytes     new bytes     delta
BenchmarkConvertToAPIRegions-12     2162007       1947634       -9.92%
```

**`h.rd.JSON` optimization**

- added `server/api/router.go#createRender` function for creating `*render.Render`
- removed option `IndentJSON` from the `render.New` call
- added option `StreamingJSON` to the  `render.New` call
- added `server/api/region_test.go#BenchmarkRenderJSON` for testing the performance of the last part of the `regionsHandler.GetAll` function (which takes 90% of the total time for the `/pd/api/v1/regions` endpoint in PCP #1837)

```sh
# GO111MODULE=on go test github.com/pingcap/pd/server/api -run=Benchmark -bench=BenchmarkRenderJSON -benchmem > new.txt
➜ benchcmp old.txt new.txt
benchmark                  old ns/op     new ns/op     delta
BenchmarkRenderJSON-12     662659672     147632381     -77.72%

benchmark                  old allocs    new allocs    delta
BenchmarkRenderJSON-12     50            12            -76.00%

benchmark                  old bytes     new bytes     delta
BenchmarkRenderJSON-12     414771916     51753928      -87.52%
```

### Check List <!--REMOVE the items that are not applicable-->

Tests <!-- At least one of them must be included. -->

 - Unit test
 - Integration test

nolouch commented:
Thanks, @ekalinin, Greatly work.
In the origin issue, the `convertToAPIRegions` also allocated much memory, can we do something for this?
![image](https://user-images.githubusercontent.com/6428910/69596623-2f8ce100-103e-11ea-88e7-cacb092219d1.png)

ekalinin commented:
> In the origin issue, the convertToAPIRegions also allocated much memory, can we do something for this?

Sure, will take a look on it today.

nolouch commented on server/api/region_test.go:
```diff
@@ -331,3 +333,72 @@ func (s *testGetRegionSuite) TestScanRegionByKey(c *C) {
+// Create n regions (0..n) of n stores (0..n).
+// Each region contains np peers, the first peer is the leader.
+// (copied from server/cluster_test.go)
+func newTestRegions(n, np uint64) []*core.RegionInfo {
```
np does not represent the number of stores, mostly used we use 3 for it. you can think it as one data have 3 replicas.

ekalinin replied:
Done

nolouch commented on server/api/region_test.go:
```diff
@@ -331,3 +333,72 @@ func (s *testGetRegionSuite) TestScanRegionByKey(c *C) {
+		region := &metapb.Region{
+			Id:          i,
+			Peers:       peers,
+			StartKey:    []byte{byte(i)},
```
maybe use `fmt.Sprintf("%d",i)`, `i ` bigger than uint8 in your case.

ekalinin replied:
Done

nolouch commented on server/api/region_test.go:
```diff
@@ -331,3 +333,72 @@ func (s *testGetRegionSuite) TestScanRegionByKey(c *C) {
+
+func BenchmarkRenderJSON(b *testing.B) {
+	regionsCount := uint64(10000)
+	storeCount := uint64(100)
```
np := 3

ekalinin replied:
Done

rleungx commented on server/api/region.go:
```diff
@@ -50,22 +50,32 @@ func NewRegionInfo(r *core.RegionInfo) *RegionInfo {
 	if r == nil {
```
It seems we can remove it.

ekalinin replied:
Done

ekalinin commented:
Changes:

- `region_test.newTestRegions`: removed args, used predefined args
- added `Slice/String` functions for zero-cost conversations
- added `ToUpperASCIIInplace` function (zero cost copy of `ToUppper`)
- added `EncodeToString`. It overrides `hex.EncodeToString`. Difference: returns `[]byte`, not `string`


Results:

```sh
# GO111MODULE=on go test github.com/pingcap/pd/server/api -run=Benchmark -bench=BenchmarkHexRegionKey -benchmem

➜ benchcmp hex.old.txt hex.new.txt 
benchmark                       old ns/op     new ns/op     delta
BenchmarkHexRegionKey-12        532           137           -74.25%
BenchmarkHexRegionKeyStr-12     470           140           -70.21%

benchmark                       old allocs     new allocs     delta
BenchmarkHexRegionKey-12        5              1              -80.00%
BenchmarkHexRegionKeyStr-12     3              1              -66.67%

benchmark                       old bytes     new bytes     delta
BenchmarkHexRegionKey-12        400           48            -88.00%
BenchmarkHexRegionKeyStr-12     240           48            -80.00%
```

and in general:

```sh
➜ benchcmp old.api.txt new.api.txt
benchmark                           old ns/op     new ns/op     delta
BenchmarkConvertToAPIRegions-12     2880085       1550980       -46.15%

benchmark                           old allocs     new allocs     delta
BenchmarkConvertToAPIRegions-12     82182          20003          -75.66%

benchmark                           old bytes     new bytes     delta
BenchmarkConvertToAPIRegions-12     2162007       2002431       -7.38%
```

ekalinin commented:
/run-all-tests

ekalinin commented:
> I do the test same as the original issue, with 1,000,000 regions:

Could you add a benchmark which would be close to that test with 1M regions?

> It's almost 3.5x faster now. 

Thanks! That's great!

> but 5s may still too long, can we have more method to optimize it?
> maybe we need to focus on reducing CPU usage now:

Sure, will try to dig into that direction.

> We really need IndentJSON for human readable. can we keep them both at the same time?

Unfortunately, `IndentJSON` requires a lot of additional work (`Marshal`, then `Indent`):

https://github.com/golang/go/blob/8a5af7910a9b157c02736c3e0998a587bb8511c1/src/encoding/json/encode.go#L175-L186

Why just not use external tool? Like:

```sh
$ >> ./bin/pd-ctl -u http://0.0.0.0:32833 config show all | python -m json.tool
```

or

```sh
$ >> ./bin/pd-ctl -u http://0.0.0.0:32833 config show all | jq .
```

nolouch commented:
>  Could you add a benchmark which would be close to that test with 1M regions?

I have shared my data [in here](https://drive.google.com/file/d/1G6vA002F2xXARNlMEnfh8qs_F2QNvl_l/view?usp=sharing), you can use it to construct the regions in your local benchmark.

> Why just not use an external tool? Like

`pd-ctl` is a tool that our users or DBAs often use it to diagnose problems. , we must ensure that it is easy to use. at lease, we just let fetch all regions to use `StreamingJSON`,   others don't.

ekalinin commented:
> at lease, we just let fetch all regions to use StreamingJSON, others don't.

Done.

ekalinin commented:
> how about have a try https://github.com/json-iterator/go to getting all regions?*

Just tried it. Looks not well:

```sh
# current branch

➜ GO111MODULE=on go test github.com/pingcap/pd/server/api -run=Benchmark -bench=BenchmarkRenderJSON
goos: linux
goarch: amd64
pkg: github.com/pingcap/pd/server/api
BenchmarkRenderJSON-12    	      96	  15344540 ns/op	 2429223 B/op	       5 allocs/op
PASS
ok  	github.com/pingcap/pd/server/api	1.732s

# current branch with github.com/json-iterator/go

GO111MODULE=on go test github.com/pingcap/pd/server/api -run=Benchmark -bench=BenchmarkRenderJSON
goos: linux
goarch: amd64
pkg: github.com/pingcap/pd/server/api
BenchmarkRenderJSON-12    	      27	  38831651 ns/op	 6267826 B/op	      18 allocs/op
PASS
ok  	github.com/pingcap/pd/server/api	1.328s
```

patch:

```go
func (ri *RegionsInfo) MarshalJSON() ([]byte, error) {
	var json = jsoniter.ConfigFastest
	return json.Marshal(*ri)
}
```

ekalinin commented:
Tried `github.com/bet365/jingo` as well. No luck again:

```
➜ GO111MODULE=on go test github.com/pingcap/pd/server/api -run=Benchmark -bench=BenchmarkRenderJSON
goos: linux
goarch: amd64
pkg: github.com/pingcap/pd/server/api
BenchmarkRenderJSON-12    	      15	  77177199 ns/op	17316683 B/op	   50040 allocs/op
PASS
ok  	github.com/pingcap/pd/server/api	1.426s
```

patch:

```go

var enc = jingo.NewStructEncoder(RegionInfo{})

func (ri *RegionInfo) MarshalJSON() ([]byte, error) {
 	buf := jingo.NewBufferFromPool()
 	enc.Marshal(ri, buf)

 	b := make([]byte, len(buf.Bytes))
 	copy(b, buf.Bytes)
 	buf.ReturnToPool()
 	return b, nil
}
```

nolouch commented on server/api/router.go:
```diff
@@ -98,7 +108,8 @@ func createRouter(prefix string, svr *server.Server) *mux.Router {
 
-	regionsHandler := newRegionsHandler(svr, rd)
+	srd := createStreamingRender()
+	regionsHandler := newRegionsHandler(svr, srd)
```
`regionshandler` also need `IndentRender`, we can use `StreamingRender` only for `GetAll`.

ekalinin replied:
Done

ekalinin commented:
Current results of the json render optimization:

```sh

# GO111MODULE=on go test github.com/pingcap/pd/server/api -run=Benchmark -bench=BenchmarkRenderJSON -benchmem

➜ benchcmp old.txt new.txt        
benchmark                  old ns/op     new ns/op     delta
BenchmarkRenderJSON-12     662659672     17829821      -97.31%

benchmark                  old allocs    new allocs    delta
BenchmarkRenderJSON-12     50            4             -92.00%

benchmark                  old bytes     new bytes     delta
BenchmarkRenderJSON-12     414771916     2339140       -99.44%
```

nolouch commented:
PTAL @rleungx
